script/train_entered.lua

```diff
@@ -34,141 +49,157 @@ local function create_temp_train(event,position)
 				end
 			elseif (event.cause.train.carriages[i].type == "fluid-wagon") then
-				global.TrainsInTunnel[SpookyGhost.unit_number].carriages[i].type = "fluid-wagon"
-				global.TrainsInTunnel[SpookyGhost.unit_number].carriages[i].fluids = event.cause.train.carriages[i].get_fluid_contents()
+				TrainInTunnel.carriages[i].type = "fluid-wagon"
+				TrainInTunnel.carriages[i].fluids = event.cause.train.carriages[i].get_fluid_contents()
 			end
 		end
 	end
 end
 
-local function entity_damaged(event)
-	if event.cause and event.entity then
-		object = event.cause.type
-		objectN = event.cause.name
-	else
-		object = nil
-		objectN = nil
+local function get_orientation(Exit,Entrance)
+	x = global.Tunnels[Exit].tunnel.position.x - global.Tunnels[Entrance].tunnel.position.x
+	y = global.Tunnels[Exit].tunnel.position.y - global.Tunnels[Entrance].tunnel.position.y
+	res = (math.atan2(y, x)+(math.pi/2)) / (math.pi*2)
+	if res > 2*math.pi then
+		res = res - 2*math.pi
+	end
+	return res
+end
+
+local function get_uarea(tunnel)
+	if global.Tunnels[tunnel].tunnel.orientation == 0 then return  {x=0,y=1}
+	elseif global.Tunnels[tunnel].tunnel.orientation == 0.25 then return  {x=-1,y=0}
+	elseif global.Tunnels[tunnel].tunnel.orientation == 0.50 then return  {x=0,y=-1}
+	elseif global.Tunnels[tunnel].tunnel.orientation == 0.75 then return  {x=1,y=0}
 	end
-	res, uarea = collision_check(event,1)
+end
 
-	--train entering tunnel
-	if res and object == "locomotive" then 
-		range = 5.5
-		area = math2d.position.multiply_scalar(uarea,range)
-		position = math2d.position.add(event.entity.position,area) --position for temp objects
-
-		--create proper ghostcar to bump other side
-		GhostCar =  create_ghost_car(event,position)
-		GC_unit_number = GhostCar.unit_number
-		global.TrainsInTunnel[GC_unit_number] = {}
-		global.TrainsInTunnel[GC_unit_number]["arrived"] = {}
-		global.TrainsInTunnel[GC_unit_number]["arrived"].check = false
-		global.TrainsInTunnel[SpookyGhost.unit_number].guideCar = GhostCar
-
-		--create ghost train to save the LTN schedule
-		TempTrain = create_temp_train(event,position)
-		global.TrainsInTunnel[GC_unit_number].TempTrain  = TempTrain
-		
-		
-		--copy train information
-		copy_train(event,GC_unit_number)
-		
-		--transfer passenger to ghost car
-		if (event.cause.get_driver() ~= nil) then
-			global.TrainsInTunnel[GC_unit_number].passenger = event.cause.get_driver()
-			SpookyGhost.set_passenger(event.cause.get_driver())
+local function create_ghost_car(event,position,orientation)
+	SpookyGhost = event.entity.surface.create_entity
+		({
+			name = "ghostCar",
+			position = position,
+			force = event.cause.force,
+		})
+
+	SpookyGhost.orientation = orientation
+	SpookyGhost.operable = false
+	SpookyGhost.speed = constants.GHOST_SPEED--event.cause.speed
+	SpookyGhost.destructible = false
+
+	return SpookyGhost
+end
+
+local function train_entered(event,uarea,TrainInTunnel,Exit,Entrance)
+	range = 5.5
```

Comment:
Is this range variable the collision detection range? Might be better to move to constants